pkg/hnsw/vector.go

```diff
@@ -0,0 +1,122 @@
+package hnsw
+
+import (
+	"errors"
+	"math"
+)
+
+type Point []float32
+
+type Vector struct {
+	id    Id
+	point Point
+
+	friends []*BaseQueue
+}
+
+// NewVector creates a new vector, note the max level is inclusive.
+func NewVector(id Id, point Point, maxLevel int) *Vector {
+	friends := make([]*BaseQueue, 0)
+
+	for i := 0; i < maxLevel; i++ {
+		bq := NewBaseQueue(MinComparator{})
+		friends = append(friends, bq)
+	}
+
+	return &Vector{
+		id:      id,
+		point:   point,
+		friends: friends,
+	}
+}
+
+func (v *Vector) Levels() int {
+	return len(v.friends)
+}
+
+func (v *Vector) MaxLevel() int {
+	return len(v.friends) - 1
+}
+
+func (v *Vector) HasLevel(level int) bool {
+	if level < 0 {
+		panic("level must be nonzero positive integer")
+	}
+
+	return level <= v.MaxLevel()
+}
+
+// InsertFriendsAtLevel requires level must be zero-indexed
+func (v *Vector) InsertFriendsAtLevel(level int, friend *Vector) {
+	if !v.HasLevel(level) {
+		panic("failed to insert friends at level, as level is not valId")
+	}
+
+	if friend.id == v.id {
+		panic("cannot insert yourself to friends list")
+	}
+
+	dist := v.EuclidDistance(friend)
```

Comment:
Specifically here, maybe make this insert API something like

```
func (v *Vector) Insert(level int, friend Point, dist float64)
```

This would make the distance functions more state-less.

Reply:
Done